File: desktop/src/shared/fuzzy-match.ts
/**
 * Fuzzy matching utilities for slash-command filtering.
 *
 * Pure functions — no React, no DOM dependencies.
 * Algorithm mirrors `ios/IonRemote/Utilities/FuzzyMatch.swift` for
 * cross-platform parity.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface FuzzyCandidate {
  command: string
  group?: string
}

export interface FuzzyMatchResult {
  match: boolean
  score: number
}

// ---------------------------------------------------------------------------
// Scoring constants
// ---------------------------------------------------------------------------

const BONUS_EXACT_PREFIX = 20
const BONUS_START_OF_STRING = 10
const BONUS_SEGMENT_BOUNDARY = 8
const BONUS_CONSECUTIVE = 5

const SEPARATORS = new Set(['-', '_', ':'])

const GROUP_ORDER: Record<string, number> = {
  builtin: 0,
  project: 1,
  extension: 2,
  user: 3,
}

// ---------------------------------------------------------------------------
// fuzzyMatchCommand
// ---------------------------------------------------------------------------

/**
 * Attempt a fuzzy subsequence match of `query` against `commandName`.
 *
 * Leading `/` is stripped from both query and command name before matching
 * so that `/clear`, `clear`, and `/clear` vs `/clear` all work identically.
 *
 * Returns `null` when the query is not a subsequence of the candidate.
 * Otherwise returns `{ match: true, score }`.
 */
export function fuzzyMatchCommand(
  query: string,
  commandName: string,
): FuzzyMatchResult | null {
  // Normalise: strip a leading slash from both, lowercase both.
  const q = stripSlash(query).toLowerCase()
  const name = stripSlash(commandName).toLowerCase()

  // Empty query matches everything with a neutral score.
  if (q.length === 0) {
    return { match: true, score: 0 }
  }

  // --- Subsequence gate --------------------------------------------------
  // Walk through `name` consuming characters from `q` in order.
  // Along the way, record the index in `name` where each query char matched
  // so we can compute bonuses in one pass.

  const matchIndices: number[] = []
  let qi = 0

  for (let ni = 0; ni < name.length && qi < q.length; ni++) {
    if (name[ni] === q[qi]) {
      matchIndices.push(ni)
      qi++
    }
  }

  if (qi !== q.length) {
    // Not every query character was consumed — no subsequence match.
    return null
  }

  // --- Scoring -----------------------------------------------------------

  let score = 0

  for (let i = 0; i < matchIndices.length; i++) {
    const idx = matchIndices[i]

    // Base: +1 per matched character.
    score += 1

    // Start-of-string bonus (first character of the name).
    if (idx === 0) {
      score += BONUS_START_OF_STRING
    }

    // Segment-boundary bonus: character immediately after a separator.
    if (idx > 0 && SEPARATORS.has(name[idx - 1])) {
      score += BONUS_SEGMENT_BOUNDARY
    }

    // Consecutive-match bonus: current match index is exactly one past the
    // previous match index (i.e. the characters are adjacent in `name`).
    if (i > 0 && idx === matchIndices[i - 1] + 1) {
      score += BONUS_CONSECUTIVE
    }
  }

  // Exact prefix bonus: the entire query matches the start of the candidate.
  if (matchIndices.length === q.length && matchIndices[0] === 0 && matchIndices[matchIndices.length - 1] === q.length - 1) {
    score += BONUS_EXACT_PREFIX
  }

  return { match: true, score }
}

// ---------------------------------------------------------------------------
// fuzzyFilterAndSort
// ---------------------------------------------------------------------------

/**
 * Filter `commands` to those that fuzzy-match `query`, then sort by:
 *   1. Score descending (best match first)
 *   2. Group order ascending  (builtin → project → extension → user)
 *   3. Alphabetical by command name
 */
export function fuzzyFilterAndSort<T extends FuzzyCandidate>(
  query: string,
  commands: T[],
): T[] {
  const scored: Array<{ item: T; score: number }> = []

  for (const cmd of commands) {
    const result = fuzzyMatchCommand(query, cmd.command)
    if (result !== null) {
      scored.push({ item: cmd, score: result.score })
    }
  }

  scored.sort((a, b) => {
    // 1. Higher score first.
    if (a.score !== b.score) return b.score - a.score

    // 2. Group order.
    const ga = GROUP_ORDER[a.item.group || 'builtin'] ?? 99
    const gb = GROUP_ORDER[b.item.group || 'builtin'] ?? 99
    if (ga !== gb) return ga - gb

    // 3. Alphabetical.
    return a.item.command.localeCompare(b.item.command)
  })

  return scored.map((s) => s.item)
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function stripSlash(s: string): string {
  return s.startsWith('/') ? s.slice(1) : s
}

File: ios/IonRemote/Utilities/FuzzyMatch.swift
/// Fuzzy matching utility for slash command autocomplete.
/// Algorithm mirrors `desktop/src/shared/fuzzy-match.ts` for cross-platform parity.
enum FuzzyMatch {

    // MARK: - Separators

    private static let separators: Set<Character> = ["-", "_", ":"]

    // MARK: - Public API

    /// Returns `nil` if the query is not a subsequence of the candidate.
    /// Otherwise returns the fuzzy score (higher = better match).
    ///
    /// Scoring:
    /// - Exact prefix match: **+20** bonus (once, if entire query matches start of candidate)
    /// - Match at start of string: **+10** bonus per character
    /// - Match after separator (`-`, `_`, `:`): **+8** bonus per character
    /// - Consecutive matches: **+5** bonus per consecutive character after the first
    /// - Base: **1** point per matched character
    static func score(query: String, candidate: String) -> Int? {
        // 1. Strip leading "/" and lowercase both strings.
        let q = stripped(query)
        let c = stripped(candidate)

        // 2. Empty query matches everything with score 0.
        if q.isEmpty { return 0 }

        // 3. Subsequence gate – every character in q must appear in order in c.
        //    While checking, record the matching indices for scoring.
        let cChars = Array(c)
        let qChars = Array(q)

        var matchIndices = [Int]()
        matchIndices.reserveCapacity(qChars.count)

        var ci = 0
        for qChar in qChars {
            var found = false
            while ci < cChars.count {
                if cChars[ci] == qChar {
                    matchIndices.append(ci)
                    ci += 1
                    found = true
                    break
                }
                ci += 1
            }
            if !found {
                return nil  // not a subsequence
            }
        }

        // 4. Compute score from the matched indices.
        var total = 0

        for (i, matchIndex) in matchIndices.enumerated() {
            // Base: +1 per matched character
            total += 1

            // Start-of-string bonus: +10 per character matched at index 0
            if matchIndex == 0 {
                total += 10
            }

            // After-separator bonus: +8 per character immediately following a separator
            if matchIndex > 0 && separators.contains(cChars[matchIndex - 1]) {
                total += 8
            }

            // Consecutive bonus: +5 for each character that is consecutive with the
            // previous match (only the 2nd, 3rd, … in a streak get the bonus).
            if i > 0 && matchIndex == matchIndices[i - 1] + 1 {
                total += 5
            }
        }

        // Exact prefix bonus: +20 if the entire query matches the start of the candidate.
        if matchIndices.count == qChars.count
            && matchIndices.first == 0
            && matchIndices.last == qChars.count - 1 {
            total += 20
        }

        return total
    }

    // MARK: - Helpers

    /// Strip a leading `/` and lowercase the string.
    private static func stripped(_ s: String) -> String {
        var result = s
        if result.hasPrefix("/") {
            result = String(result.dropFirst())
        }
        return result.lowercased()
    }
}
